### PR TITLE
Convert boxel-ui to v2 addon

### PR DESCRIPTION
**Pull request by lukemelia**
- boxel-ui is now built with rollup: `cd packages/boxel-ui/addon && pnpm build`
- boxel-ui has four modules with external exports:
  - @cardstack/boxel-ui/components (`import { BoxelButton } from '@cardstack/boxel-ui/components';`)
  - @cardstack/boxel-ui/helpers (`import { cn } from '@cardstack/boxel-ui/helpers';`)
  - @cardstack/boxel-ui/icons (`import { Warning } from '@cardstack/boxel-ui/icons';`)
  - @cardstack/boxel-ui/usage (`import { BoxelButtonUsage } from '@cardstack/boxel-ui/usage';`)
- svgjar is no longer used
- Freestyle app now lives in boxel-ui/test-app
- imports _within_ the addon must specify extension (.ts or .gts)
- imports from outside the addon do not include an extension
- boxel-ui/addon's package.json include `rebuild:icons` and `rebuild:usage` to code generate icons and exports

### CODE REVIEW

**Comment by backspace** [comment]
I’ll look more thoroughly but early feedback: should `packages/boxel-ui/.ember-cli` be moved into `…/test-app`? Then we get port 4210 by default when running it.

**Comment by lukemelia** [comment]
> I’ll look more thoroughly but early feedback: should packages/boxel-ui/.ember-cli be moved into …/test-app? Then we get port 4210 by default when running it.

Yep. Updated.

File: packages/boxel-ui/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Note: icon components in addon/src/icons are code-generated from svg files using `pnpm rebuild:icons` from the addon project.
```
Comment (backspace) [inline packages/boxel-ui/README.md]: Should this explain how to run with Freestyle? And that you need to `pnpm start` in `…/addon` for that to work?

File: package.json [hunk 2]
```diff
@@ -5,11 +5,11 @@
...
-    "deploy:boxel-host:preview-staging": "cd packages/host && BASE_REALM_HOSTING_DISABLED=true NODE_OPTIONS='--max_old_space_size=4096' pnpm exec ember deploy s3-preview-staging --verbose",
-    "deploy:boxel-host:preview-production": "cd packages/host && BASE_REALM_HOSTING_DISABLED=true NODE_OPTIONS='--max_old_space_size=4096' pnpm exec ember deploy s3-preview-production --verbose",
+    "deploy:boxel-host": "cd packages/boxel-ui/addon && pnpm build && popd && cd packages/host && BASE_REALM_HOSTING_DISABLED=true NODE_OPTIONS='--max_old_space_size=4096' pnpm exec ember deploy",
+    "deploy:boxel-host:preview-staging": "cd packages/boxel-ui/addon && pnpm build && cd ../../host && BASE_REALM_HOSTING_DISABLED=true NODE_OPTIONS='--max_old_space_size=4096' pnpm exec ember deploy s3-preview-staging --verbose",
+    "deploy:boxel-host:preview-production": "cd packages/boxel-ui/addon && pnpm build && cd ../../host && BASE_REALM_HOSTING_DISABLED=true NODE_OPTIONS='--max_old_space_size=4096' pnpm exec ember deploy s3-preview-production --verbose",
```
Comment (backspace) [inline package.json:12]: Could these three be prefixed with `pnpm build:boxel-host && ` instead?

File: packages/host/app/helpers/truth-helpers.ts [hunk 3]
```diff
@@ -1,15 +0,0 @@
...
-}
-export function not(a: any): boolean {
-  return !a;
-}
-export function or(a: boolean, b: boolean): boolean {
-  return a || b;
-}
-export function and(a: boolean, b: boolean): boolean {
-  return a && b;
-}
```
Comment (backspace) [inline packages/host/app/helpers/truth-helpers.ts:15]: I’m glad to see this, I was confused why we had it in both places!

File: packages/boxel-ui/addon/.template-lintrc.cjs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+module.exports = {
+  extends: 'recommended',
+  rules: {
+    'require-button-type': false,
```
Comment (backspace) [inline packages/boxel-ui/addon/.template-lintrc.cjs:6]: Why is this needed? 😯

File: packages/boxel-ui/addon/src/components.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import BoxelButton from './components/button/index.gts';
+import CardContainer from './components/card-container/index.gts';
+import BoxelDropdown from './components/dropdown/index.gts';
+import DropdownButton from './components/dropdown-button/index.gts';
+import FieldContainer from './components/field-container/index.gts';
+import GridContainer from './components/grid-container/index.gts';
+import BoxelHeader from './components/header/index.gts';
+import Header from './components/header/index.gts';
+import IconButton from './components/icon-button/index.gts';
+import BoxelInput from './components/input/index.gts';
```
Comment (habdelra) [inline packages/boxel-ui/addon/src/components.ts:13]: so for this new style, does VS code still understand these imports, and is glint still giving us types within vs code for these imports?

Comment (lukemelia) [reply packages/boxel-ui/addon/src/components.ts:13]: Yep!

Comment (habdelra) [reply packages/boxel-ui/addon/src/components.ts:13]: Also, if we fail to import with the file extension, will linting warn us we are doing it wrong? (and/or are build errors are clear enough to guide us to the correct action)

File: packages/boxel-ui/addon/src/components/card-container/index.gts [hunk 6]
```diff
@@ -1,22 +1,23 @@
...
+import Header from '../header/index.gts';
 
 interface Signature {
-  Element: HTMLElement;
   Args: {
-    tag?: keyof HTMLElementTagNameMap;
+    displayBoundaries?: boolean;
+    isHighlighted?: boolean;
     label?: string;
+    tag?: keyof HTMLElementTagNameMap;
```
Comment (backspace) [inline packages/boxel-ui/addon/src/components/card-container/index.gts:13]: I like these being sorted, can we automate it? (Or maybe it is?)

File: packages/boxel-ui/addon/src/icons.gts
Comment (habdelra) [inline packages/boxel-ui/addon/src/icons.gts:1]: what is the protocol for this auto generated file. Why do we check it into git? and when should we re-run the script to rebuild this?

Comment (habdelra) [reply packages/boxel-ui/addon/src/icons.gts:1]: ditto for all the other auto generated files

File: packages/boxel-ui/addon/src/styles/fonts.css
Comment (habdelra) [inline packages/boxel-ui/addon/src/styles/fonts.css:1]: should we remove the commented out code?

File: packages/boxel-ui/addon/src/usage.ts
Comment (habdelra) [inline packages/boxel-ui/addon/src/usage.ts:1]: similar question as the rebuild icons. why are we checking this into git, and what is the protocol for when we need to re-run this script (pnpm rebuild:usage)

**Review by habdelra — COMMENTED** [review]
it looks good, although i'd like to see more documentation around the rebuild scripts, I think there are 2 of them. it can be confusing if a developer changes some code but that can't see those changes until they run a special script. ideally it would be nice if the build artifacts were never checked into git and the build would just happen automatically instead of a special step that must be taught to everyone.

**Review by habdelra — APPROVED** [review]

File: packages/boxel-ui/README.md (hunk 1 above)
Comment (habdelra) [inline packages/boxel-ui/README.md]: i think there is another rebuild script for usage right?

File: packages/boxel-ui/addon/src/icons.gts
Comment (lukemelia) [reply packages/boxel-ui/addon/src/icons.gts:1]: @ef4 and I were thinking that you would re-run the script when you add, remove, or update an icon, which is relatively infrequently. Open to other thoughts about automating or linting this process.

File: packages/boxel-ui/addon/src/usage.ts
Comment (lukemelia) [reply packages/boxel-ui/addon/src/usage.ts:1]: Same answer -- when you add, remove, or rename a usage file. I'll update docs to include this one.

File: packages/boxel-ui/README.md (hunk 1 above)
Comment (lukemelia) [reply packages/boxel-ui/README.md]: I improved these docs.

File: packages/boxel-ui/addon/src/icons.gts
Comment (habdelra) [reply packages/boxel-ui/addon/src/icons.gts:1]: perhaps we write a checksum into a comment in the auto generated file, and we could have a lint that checks the source checksum vs the checksum recorded in the auto generated file. I guess I'm just trying to save the time of the developer that does need to change one of these and forgets or doesn't realize they need to run these scripts (i've had to make perhaps 6 additions/changes in svgs and usages in the past couple months--for reference Fadhlan just added an SVG in a pending PR). it can be maddening when you change a thing but don't see it reflected in the app.

File: packages/boxel-ui/addon/src/components.ts (hunk 5 above)
Comment (lukemelia) [reply packages/boxel-ui/addon/src/components.ts:13]: > Also, if we fail to import with the file extension, will linting warn us we are doing it wrong? (and/or are build errors are clear enough to guide us to the correct action)

There is definitely room for improvement here.

File: packages/boxel-ui/addon/src/styles/fonts.css
Comment (lukemelia) [reply packages/boxel-ui/addon/src/styles/fonts.css:1]: Seems like yes. This CSS was referencing font files that did not exist, and it became an error under rollup.

File: packages/boxel-ui/README.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+Icon components in addon/src/icons are code-generated from svg files using `pnpm rebuild:icons` from the addon project. This script also generates src/icons.ts, which is the module that re-exports the icons for consumers of this addon. 
+
+This script should be run when an icon is added, removed, updated, or renamed.
+
+### `pnpm rebuild:usage`
+
+This script generates src/usage.ts, which is the module that re-exports the usage files for the test-app to include in it's component explorer UI. 
+
+This script should be run when a usage file is added, removed, or renamed.
```
Comment (habdelra) [inline packages/boxel-ui/README.md:13]: to be clear, when a file is removed or renamed the script probably isn't gonna handle cleaning up the old artifact. i think we'll need to delete that artifact by hand, no?

**Comment by backspace** [comment]
I’m still working my way through the files but when I cloned a fresh repository and did this setup:

* `pnpm install`
* `cd packages/boxel-ui/addon && pnpm rebuild:icons && pnpm rebuild:usage`

Then when I ran `pnpm start` in `packages/boxel-ui/test-app` I got this:

```
Module not found: Error: Can't resolve '@cardstack/boxel-ui/styles/global.css' in '/Users/b/Documents/Cardstack/Code/b9/packages/boxel-ui/addon'
```

Did I miss something?

What I was hoping to see with this is whether the icons that exist are listed, as this is a useful feature of SVG Jar. Browsing the icons in the filesystem doesn’t work well when they have CSS variables for colours because they end up transparent.

**Review by backspace — COMMENTED** [review]
Some of my comments may be outdated already as it took me a while to review. I’m still unable to run this locally but I may have missed a step!

**Comment by lukemelia** [comment]
@backspace wrote:
> Did I miss something?

`pnpm build` (or `pnpm start`) in `packages/boxel-ui/addon`

File: packages/boxel-ui/addon/.template-lintrc.cjs (hunk 4 above)
Comment (lukemelia) [reply packages/boxel-ui/addon/.template-lintrc.cjs:6]: Not sure. I twas not added intentionally by me. It was either here already, or was auto-generated.

**Comment by backspace** [comment]
> @backspace wrote:
> 
> > Did I miss something?
> 
> `pnpm build` in `packages/boxel-ui/addon`

lol indeed, it runs now, thanks.

re the SVG browser maybe there could be a list of icons in Freestyle?

**Comment by lukemelia** [comment]
> re the SVG browser maybe there could be a list of icons in Freestyle

This would be easy to add post-merge 👍

File: packages/boxel-ui/addon/src/components/card-container/index.gts (hunk 6 above)
Comment (lukemelia) [reply packages/boxel-ui/addon/src/components/card-container/index.gts:13]: 🤷

**Comment by lukemelia** [comment]
Merged! Let's discuss how we can address the DX issues with Ed in retro tomorrow
